{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "24934cdd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: mechanize in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (0.4.7)\n",
      "Requirement already satisfied: html5lib>=0.999999999 in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (from mechanize) (1.1)\n",
      "Requirement already satisfied: six>=1.9 in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (from html5lib>=0.999999999->mechanize) (1.15.0)\n",
      "Requirement already satisfied: webencodings in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (from html5lib>=0.999999999->mechanize) (0.5.1)\n",
      "Requirement already satisfied: Bio in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (1.3.2)\n",
      "Requirement already satisfied: biopython>=1.79 in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (from Bio) (1.79)\n",
      "Requirement already satisfied: mygene in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (from Bio) (3.2.2)\n",
      "Requirement already satisfied: requests in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (from Bio) (2.25.1)\n",
      "Requirement already satisfied: tqdm in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (from Bio) (4.59.0)\n",
      "Requirement already satisfied: numpy in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (from biopython>=1.79->Bio) (1.20.1)\n",
      "Requirement already satisfied: biothings-client>=0.2.6 in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (from mygene->Bio) (0.2.6)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (from requests->Bio) (2.10)\n",
      "Requirement already satisfied: chardet<5,>=3.0.2 in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (from requests->Bio) (4.0.0)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (from requests->Bio) (1.26.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/zoeweiss/opt/anaconda3/lib/python3.8/site-packages (from requests->Bio) (2020.12.5)\n"
     ]
    }
   ],
   "source": [
    "#IMPORTS\n",
    "#Install+Import Packages\n",
    "!pip install mechanize\n",
    "import random\n",
    "import re\n",
    "import mechanize\n",
    "import numpy as np\n",
    "from random import *\n",
    "import matplotlib.pyplot as plt\n",
    "!pip install Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from collections import Counter \n",
    "from Bio.Cluster import kcluster\n",
    "import statistics\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics import silhouette_score\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import difflib\n",
    "import seaborn as sns; sns.set_theme()\n",
    "from collections import defaultdict\n",
    "import collections\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "from pylab import *\n",
    "from collections import OrderedDict\n",
    "import matplotlib.pyplot as plt\n",
    "from itertools import filterfalse\n",
    "from matplotlib.pyplot import figure\n",
    "import scipy.stats as stats\n",
    "import pylab\n",
    "from itertools import permutations\n",
    "#!pip install ipython-autotime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d3c69bb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Upload Data and Define Parameters\n",
    "leu_aa = open('/Users/zoeweiss/Desktop/Desktop/lab(s)/szostak/spring 2022/aminoacylation/CP1.3/Leu_S5_L001_R1_001.fastq').readlines()\n",
    "ala_aa = open('/Users/zoeweiss/Desktop/Desktop/lab(s)/szostak/spring 2022/aminoacylation/CP1.3/Ala_S3_L001_R1_001.fastq').readlines()\n",
    "gly_aa = open('/Users/zoeweiss/Desktop/Desktop/lab(s)/szostak/spring 2022/aminoacylation/CP1.3/Gly_S1_L001_R1_001.fastq').readlines()\n",
    "phe_aa = open('/Users/zoeweiss/Desktop/Desktop/lab(s)/szostak/spring 2022/aminoacylation/CP1.3/Phe_S7_L001_R1_001.fastq').readlines()\n",
    "\n",
    "leu_rna = open('/Users/zoeweiss/Desktop/Desktop/lab(s)/szostak/spring 2022/aminoacylation/CP1.3/LeuNaOH_S6_L001_R1_001.fastq').readlines()\n",
    "ala_rna = open('/Users/zoeweiss/Desktop/Desktop/lab(s)/szostak/spring 2022/aminoacylation/CP1.3/AlaNaOH_S4_L001_R1_001.fastq').readlines()\n",
    "gly_rna = open('/Users/zoeweiss/Desktop/Desktop/lab(s)/szostak/spring 2022/aminoacylation/CP1.3/GlyNaOH_S2_L001_R1_001.fastq').readlines()\n",
    "phe_rna = open('/Users/zoeweiss/Desktop/Desktop/lab(s)/szostak/spring 2022/aminoacylation/CP1.3/PheNaOH_S8_L001_R1_001.fastq').readlines()\n",
    "\n",
    "ctrl_8a = open('/Users/zoeweiss/Desktop/Desktop/lab(s)/szostak/spring 2022/aminoacylation/CP1.3/8actrl_S9_L001_R1_001.fastq').readlines()\n",
    "ctrl_5b = open('/Users/zoeweiss/Desktop/Desktop/lab(s)/szostak/spring 2022/aminoacylation/CP1.3/5bctrl_S10_L001_R1_001.fastq').readlines()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7c64fda2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Upload Data and Define Parameters\n",
    "file_type = '.fastq'\n",
    "\n",
    "#Take the reverse complement? Enter 'yes' or 'no'\n",
    "rc = 'no'\n",
    "\n",
    "#Quality Cutoff - what fraction of positions should have <1% error probability?\n",
    "quality_cutoff = 0.9\n",
    "\n",
    "\n",
    "#Position motif1 starts at\n",
    "motif_start = 1\n",
    "#Motif1 define\n",
    "motif = 'AAAGCCTG'\n",
    "#Position motif2 starts at\n",
    "motif2_start = 21\n",
    "#Motif2 define\n",
    "motif2 = 'TAGGCAAAAGATCGGAAGAGCACACGTCT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f38104ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_seqs(list_of_seqs):\n",
    "\n",
    "    if file_type == '.fastq':\n",
    "        seqs = list_of_seqs[1::4]\n",
    "        qualities = list_of_seqs[3::4]\n",
    "        print('Total seqs: '+str(len(seqs)))\n",
    "\n",
    "        #Filter by quality\n",
    "        high_quality = []\n",
    "        acceptable_quals_1_percent_error = [5,6,7,8,9,':',';', '<', '=', '>', '?', '@', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K']\n",
    "        for seq in range(len(qualities)):\n",
    "            overlap=0\n",
    "            for char in list(set(qualities[seq]).intersection(acceptable_quals_1_percent_error)):\n",
    "                overlap+=qualities[seq].count(char)\n",
    "            if overlap/len(qualities[1])>quality_cutoff:\n",
    "                high_quality.append(seqs[seq].strip(\"\\n\"))\n",
    "        print('High quality seqs: '+str(len(high_quality)))\n",
    "    \n",
    "    if file_type == '.fasta':\n",
    "        high_quality = list_of_seqs[1::2]\n",
    "        print('Total seqs: '+str(len(high_quality)))\n",
    "    \n",
    "    #Take RC if desired\n",
    "    pre_trim = []\n",
    "    if rc == 'yes':\n",
    "        rc_seqs = []\n",
    "        for sequence in high_quality:\n",
    "            rc_seqs.append(str(Seq(sequence).reverse_complement()).strip('\\n'))\n",
    "        pre_trim = rc_seqs\n",
    "    else:\n",
    "        for sequence in high_quality:\n",
    "            pre_trim.append(str(Seq(sequence)).strip('\\n'))\n",
    "    if rc == 'yes':\n",
    "        print('Reverse Comp = '+str(rc)) \n",
    "\n",
    "    #Trim to keep only seqs with constant region\n",
    "    trimmed = []\n",
    "    for seq in pre_trim:\n",
    "        if motif in seq:\n",
    "            if motif2 in seq:\n",
    "                trimmed.append('GCCTG'+seq[(motif_start-1)+len(motif):(motif2_start-1)]+'TAGGC')\n",
    "    print('Seqs after trimming: '+str(len(trimmed))) \n",
    "    \n",
    "    #All processed seqs\n",
    "    processed_seqs = trimmed\n",
    "    #Unique seqs\n",
    "    processed_seqs_unique = list(set(trimmed))\n",
    "    \n",
    "    return processed_seqs, processed_seqs_unique\n",
    "\n",
    "\n",
    "def rscape_format(title, num_seqs_to_analyze, top_seqs):\n",
    "    #Create file for R-Scape Covariation\n",
    "    f = open(\"/Users/zoeweiss/Desktop/Desktop/lab(s)/szostak/spring 2022/aminoacylation/\"+str(title)+\".txt\", \"w\")\n",
    "    f.write('# STOCKHOLM 1.0\\n')\n",
    "    f.write('\\n')\n",
    "    counter = 0\n",
    "    for seq in top_seqs[:num_seqs_to_analyze]:\n",
    "      counter+=1\n",
    "      f.write(str(counter)+'\\t'+seq[0]+'\\n')\n",
    "    #f.write('#=GC SS_cons\t(((((((((,)))))))))\\n')\n",
    "    f.write('//')\n",
    "\n",
    "def get_info(aa, rna, title):\n",
    "    print(title)\n",
    "    print('AA ONLY:')\n",
    "    aa_ctrl_processed_seqs, aa_ctrl_processed_seqs_unique = preprocess_seqs(aa)\n",
    "    aa_ctrl_top_seqs = Counter(aa_ctrl_processed_seqs).most_common(100)\n",
    "    print('Unique AA seqs '+str(len(aa_ctrl_processed_seqs_unique)))\n",
    "    print(aa_ctrl_top_seqs[:10])\n",
    "    print('RNA ONLY:')\n",
    "    rna_ctrl_processed_seqs, rna_ctrl_processed_seqs_unique = preprocess_seqs(rna)\n",
    "    rna_ctrl_top_seqs = Counter(rna_ctrl_processed_seqs).most_common(100)\n",
    "    print('Unique RNA seqs '+str(len(rna_ctrl_processed_seqs_unique)))\n",
    "    print(rna_ctrl_top_seqs[:10])\n",
    "    aa_rna_no_overlap = list(filterfalse(set(rna_ctrl_processed_seqs).__contains__, aa_ctrl_processed_seqs))\n",
    "    print('AA NOT RNA: '+str(round(len(aa_rna_no_overlap))))\n",
    "    aa_not_rna_top_seqs = Counter(aa_rna_no_overlap).most_common(10000)\n",
    "    print('Unique AA only seqs '+str(len(set(aa_rna_no_overlap))))\n",
    "    top_seqs = aa_not_rna_top_seqs[:10]\n",
    "    rscape_format(title+'_rscape_1000',1000, aa_not_rna_top_seqs)\n",
    "    print('R-Scape file saved')\n",
    "    return aa_ctrl_processed_seqs, rna_ctrl_processed_seqs, aa_rna_no_overlap, aa_ctrl_top_seqs, rna_ctrl_top_seqs, aa_not_rna_top_seqs, top_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d8bebfcf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gly\n",
      "AA ONLY:\n",
      "Total seqs: 356811\n",
      "High quality seqs: 347603\n",
      "Seqs after trimming: 216113\n",
      "Unique AA seqs 153628\n",
      "[('GCCTGGGGGGTAGGCAATAGGC', 31), ('GCCTGGGTGGTAGGCAATAGGC', 29), ('GCCTGTTATGGGAGTAATAGGC', 28), ('GCCTGGGGGTAGGCAAATAGGC', 27), ('GCCTGATGTGGGAATGTTAGGC', 26), ('GCCTGTTTGTAGGTAGATAGGC', 26), ('GCCTGATAGTGGGGTGTTAGGC', 24), ('GCCTGATTGTAGGGGGTTAGGC', 24), ('GCCTGGATTGGGAGGTTTAGGC', 24), ('GCCTGAATGTGGGGGTTTAGGC', 24)]\n",
      "RNA ONLY:\n",
      "Total seqs: 574755\n",
      "High quality seqs: 551138\n",
      "Seqs after trimming: 332641\n",
      "Unique RNA seqs 211879\n",
      "[('GCCTGGTGGTAGGCAAATAGGC', 784), ('GCCTGGTAGGCAAAAGATAGGC', 739), ('GCCTGGTATTAGGCAAATAGGC', 609), ('GCCTGGGTGGTAGGCAATAGGC', 575), ('GCCTGGGGGGTAGGCAATAGGC', 525), ('GCCTGGGGGGGTAGGCATAGGC', 514), ('GCCTGGGGGGTTAGGCATAGGC', 505), ('GCCTGGGGGTAGGCAAATAGGC', 468), ('GCCTGGTGTTAGGCAAATAGGC', 449), ('GCCTGGTGTAGGCAAAATAGGC', 440)]\n",
      "AA NOT RNA: 144422\n",
      "Unique AA only seqs 118777\n",
      "R-Scape file saved\n",
      "\n",
      "Leu\n",
      "AA ONLY:\n",
      "Total seqs: 239837\n",
      "High quality seqs: 233354\n",
      "Seqs after trimming: 181652\n",
      "Unique AA seqs 115540\n",
      "[('GCCTGATGTGGGAATGTTAGGC', 47), ('GCCTGAAATGTGAGTTTTAGGC', 42), ('GCCTGATGTGGAAATGTTAGGC', 41), ('GCCTGAAATGGGAGTTTTAGGC', 38), ('GCCTGTTAAGGGATTAATAGGC', 37), ('GCCTGAGTAGTGATATTTAGGC', 37), ('GCCTGGAATGGGAGTTTTAGGC', 37), ('GCCTGTTATGTGAATAGTAGGC', 37), ('GCCTGGTTGGTGAGGACTAGGC', 37), ('GCCTGTGTTGTGAAATGTAGGC', 36)]\n",
      "RNA ONLY:\n",
      "Total seqs: 721611\n",
      "High quality seqs: 701156\n",
      "Seqs after trimming: 465799\n",
      "Unique RNA seqs 270490\n",
      "[('GCCTGCGTGACGTGTGGTAGGC', 296), ('GCCTGCGTGGACGTGGGTAGGC', 273), ('GCCTGTGCTACGTGTGGTAGGC', 260), ('GCCTGCGTTGACGTGGGTAGGC', 251), ('GCCTGGTTTGCCTGGGTTAGGC', 211), ('GCCTGGTTGGACGTGGCTAGGC', 201), ('GCCTGGTGGGACGTGGCTAGGC', 191), ('GCCTGGTGTGACGTGGCTAGGC', 168), ('GCCTGGTAGGACGTGGCTAGGC', 155), ('GCCTGGTTGACGTGTGGTAGGC', 154)]\n",
      "AA NOT RNA: 81472\n",
      "Unique AA only seqs 72174\n",
      "R-Scape file saved\n",
      "\n",
      "Phe\n",
      "AA ONLY:\n",
      "Total seqs: 381247\n",
      "High quality seqs: 370630\n",
      "Seqs after trimming: 245012\n",
      "Unique AA seqs 149398\n",
      "[('GCCTGTTTTGTGAGGAATAGGC', 70), ('GCCTGATTTGTGAGGGTTAGGC', 68), ('GCCTGAATTGGGAGGTTTAGGC', 64), ('GCCTGTGTTGTGAGATATAGGC', 61), ('GCCTGAGTTGTGAGATTTAGGC', 61), ('GCCTGAATTGTGAGGTTTAGGC', 60), ('GCCTGTTTTGGGAGAAATAGGC', 59), ('GCCTGTTTTGTGAGGGATAGGC', 59), ('GCCTGTTATGTGAGTAGTAGGC', 58), ('GCCTGATTTGTGAGAGTTAGGC', 57)]\n",
      "RNA ONLY:\n",
      "Total seqs: 450933\n",
      "High quality seqs: 438167\n",
      "Seqs after trimming: 318918\n",
      "Unique RNA seqs 203946\n",
      "[('GCCTGCGTGACGTGTGGTAGGC', 358), ('GCCTGCGTGGACGTGGGTAGGC', 251), ('GCCTGTGCTACGTGTGGTAGGC', 213), ('GCCTGCGTTGACGTGGGTAGGC', 190), ('GCCTGGTTGGACGTGGCTAGGC', 187), ('GCCTGCGTGGACGTGAGTAGGC', 179), ('GCCTGGTGGGACGTGGCTAGGC', 176), ('GCCTGTTTGGACGTGGATAGGC', 167), ('GCCTGGGGGACGTGTGGTAGGC', 162), ('GCCTGGTTTGCCTGGGTTAGGC', 149)]\n",
      "AA NOT RNA: 142368\n",
      "Unique AA only seqs 111346\n",
      "R-Scape file saved\n",
      "\n",
      "Ala\n",
      "AA ONLY:\n",
      "Total seqs: 278245\n",
      "High quality seqs: 270275\n",
      "Seqs after trimming: 190258\n",
      "Unique AA seqs 125278\n",
      "[('GCCTGAAATGGGAGTTTTAGGC', 51), ('GCCTGGAATGGAAGTTTTAGGC', 50), ('GCCTGATGTGGGAATGTTAGGC', 48), ('GCCTGGAATGTGAGTTTTAGGC', 47), ('GCCTGTTTTGTGAGGAATAGGC', 47), ('GCCTGAGTTGGGAGATTTAGGC', 46), ('GCCTGATTTGTGAGGGTTAGGC', 45), ('GCCTGTTTTGTGAGAAATAGGC', 44), ('GCCTGAGTTGTGAGATTTAGGC', 43), ('GCCTGGAATGGGAGTTTTAGGC', 43)]\n",
      "RNA ONLY:\n",
      "Total seqs: 652770\n",
      "High quality seqs: 633803\n",
      "Seqs after trimming: 466954\n",
      "Unique RNA seqs 279630\n",
      "[('GCCTGCGTGACGTGTGGTAGGC', 402), ('GCCTGCGTGGACGTGGGTAGGC', 341), ('GCCTGCGTTGACGTGGGTAGGC', 308), ('GCCTGGTTGGACGTGGCTAGGC', 265), ('GCCTGTGCTACGTGTGGTAGGC', 259), ('GCCTGGTGGGACGTGGCTAGGC', 257), ('GCCTGTTTGGACGTGGATAGGC', 244), ('GCCTGTGTGACGTGTGATAGGC', 226), ('GCCTGGGGGACGTGTGGTAGGC', 215), ('GCCTGGTTGACGTGTGGTAGGC', 212)]\n",
      "AA NOT RNA: 105261\n",
      "Unique AA only seqs 87049\n",
      "R-Scape file saved\n"
     ]
    }
   ],
   "source": [
    "gly_aa_ctrl_processed_seqs, gly_rna_ctrl_processed_seqs, gly_aa_rna_no_overlap, gly_aa_ctrl_top_seqs, gly_rna_ctrl_top_seqs, gly_aa_not_rna_top_seqs, gly_top_seqs = get_info(gly_aa, gly_rna, 'Gly')\n",
    "print('')\n",
    "leu_aa_ctrl_processed_seqs, leu_rna_ctrl_processed_seqs, leu_aa_rna_no_overlap, leu_aa_ctrl_top_seqs, leu_rna_ctrl_top_seqs, leu_aa_not_rna_top_seqs, leu_top_seqs = get_info(leu_aa, leu_rna, 'Leu')\n",
    "print('')\n",
    "phe_aa_ctrl_processed_seqs, phe_rna_ctrl_processed_seqs, phe_aa_rna_no_overlap, phe_aa_ctrl_top_seqs, phe_rna_ctrl_top_seqs, phe_aa_not_rna_top_seqs, phe_top_seqs = get_info(phe_aa, phe_rna, 'Phe')\n",
    "print('')\n",
    "ala_aa_ctrl_processed_seqs, ala_rna_ctrl_processed_seqs, ala_aa_rna_no_overlap, ala_aa_ctrl_top_seqs, ala_rna_ctrl_top_seqs, ala_aa_not_rna_top_seqs, ala_top_seqs = get_info(ala_aa, ala_rna, 'Ala')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "71ae6a8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 6, expected reads of any subseq: 155.42\n",
      "Second 6, expected reads of any subseq: 201.08\n",
      "AA1\n",
      "1000/10000\n",
      "2000/10000\n",
      "3000/10000\n",
      "4000/10000\n",
      "5000/10000\n",
      "6000/10000\n",
      "7000/10000\n",
      "8000/10000\n",
      "9000/10000\n",
      "10000/10000\n",
      "[['GCCTGTATCCGAGTATATAGGC', 2, 0.14, 14.285714285714285], ['GCCTGTTCTATCTGAACTAGGC', 2, 0.15, 13.333333333333334], ['GCCTGTAATGCATATTATAGGC', 5, 0.42, 11.904761904761905], ['GCCTGGATCTCTCTGATTAGGC', 2, 0.17, 11.76470588235294], ['GCCTGTACGATACGTTATAGGC', 3, 0.28, 10.714285714285714], ['GCCTGAATCGAGAGAATTAGGC', 2, 0.2, 10.0], ['GCCTGAATTTCGCAATTTAGGC', 3, 0.31, 9.67741935483871], ['GCCTGAGAGTGGAACCTTAGGC', 2, 0.22, 9.090909090909092], ['GCCTGTAAAGTCATTTATAGGC', 4, 0.44, 9.090909090909092], ['GCCTGAATATTGTAATTTAGGC', 3, 0.33, 9.09090909090909]]\n",
      "AA2\n",
      "1000/10000\n",
      "2000/10000\n",
      "3000/10000\n",
      "4000/10000\n",
      "5000/10000\n",
      "6000/10000\n",
      "7000/10000\n",
      "8000/10000\n",
      "9000/10000\n",
      "10000/10000\n",
      "[['GCCTGCTCCTCCCCAGGTAGGC', 1, 0.02, 50.0], ['GCCTGCGTTCCCCCGCGTAGGC', 1, 0.03, 33.333333333333336], ['GCCTGGACAACCACCTCTAGGC', 1, 0.03, 33.333333333333336], ['GCCTGCCCAGGCTTTTATAGGC', 2, 0.07, 28.57142857142857], ['GCCTGCCAGGCTTTTAGTAGGC', 5, 0.2, 25.0], ['GCCTGTGAAGTAACACATAGGC', 2, 0.09, 22.22222222222222], ['GCCTGGTCCTACCCGCTTAGGC', 1, 0.05, 20.0], ['GCCTGTCCGCCTATTCTTAGGC', 1, 0.05, 20.0], ['GCCTGCCCGTTTCCGCGTAGGC', 1, 0.05, 20.0], ['GCCTGTACCCTGAAGTATAGGC', 3, 0.15, 20.0]]\n",
      "AA3\n",
      "1000/10000\n",
      "2000/10000\n",
      "3000/10000\n",
      "4000/10000\n",
      "5000/10000\n",
      "6000/10000\n",
      "7000/10000\n",
      "8000/10000\n",
      "9000/10000\n",
      "10000/10000\n",
      "[['GCCTGAATGGAAATATTTAGGC', 16, 0.56, 28.57142857142857], ['GCCTGAATTGCAATATTTAGGC', 4, 0.17, 23.52941176470588], ['GCCTGTGTAGTGATCCATAGGC', 4, 0.21, 19.047619047619047], ['GCCTGAGACGTCAGTTTTAGGC', 7, 0.4, 17.5], ['GCCTGTGGAGTAATACATAGGC', 6, 0.35, 17.142857142857142], ['GCCTGAGACGGAAGTTTTAGGC', 12, 0.71, 16.901408450704228], ['GCCTGAATAGTAAGATTTAGGC', 9, 0.55, 16.363636363636363], ['GCCTGTAAAGTCATTTATAGGC', 7, 0.44, 15.909090909090908], ['GCCTGTAAAGTGAATTATAGGC', 9, 0.59, 15.254237288135593], ['GCCTGGAAAGAGATTTTTAGGC', 27, 1.79, 15.083798882681563]]\n",
      "AA4\n",
      "1000/10000\n",
      "2000/10000\n",
      "3000/10000\n",
      "4000/10000\n",
      "5000/10000\n",
      "6000/10000\n",
      "7000/10000\n",
      "8000/10000\n",
      "9000/10000\n",
      "10000/10000\n",
      "[['GCCTGTGTCCTACCGCATAGGC', 2, 0.05, 40.0], ['GCCTGTGTCCAAGAACATAGGC', 2, 0.07, 28.57142857142857], ['GCCTGCCCCCGTAGGCATAGGC', 2, 0.09, 22.22222222222222], ['GCCTGCCCCCTTAGGCATAGGC', 2, 0.09, 22.22222222222222], ['GCCTGCCCAGGTAGGCATAGGC', 11, 0.53, 20.754716981132074], ['GCCTGAATTGCCAAATTTAGGC', 3, 0.17, 17.64705882352941], ['GCCTGACTCCGGAAAGTTAGGC', 2, 0.14, 14.285714285714285], ['GCCTGAGACACCCGTGTTAGGC', 2, 0.14, 14.285714285714285], ['GCCTGAATTATGAGGTTTAGGC', 13, 0.95, 13.68421052631579], ['GCCTGTAAAGTTATTTATAGGC', 9, 0.7, 12.857142857142858]]\n"
     ]
    }
   ],
   "source": [
    "#UPDATES 4/2/22\n",
    "#Make control datasets\n",
    "ctrl_8a_subseqs = []\n",
    "for sub_seq in ctrl_8a[1::4]:\n",
    "    ctrl_8a_subseqs.append(sub_seq[8:14])\n",
    "expected_8a = len(ctrl_8a_subseqs)*0.25**6\n",
    "print('First 6, expected reads of any subseq: '+str(round(expected_8a,2)))\n",
    "\n",
    "ctrl_5b_subseqs = []\n",
    "for sub_seq in ctrl_5b[1::4]:\n",
    "    ctrl_5b_subseqs.append(sub_seq[3:9])\n",
    "expected_5b = len(ctrl_5b_subseqs)*0.25**6\n",
    "print('Second 6, expected reads of any subseq: '+str(round(expected_5b,2)))\n",
    "\n",
    "#Find enrichment for each sequence\n",
    "aa_not_rna_top_seqs = gly_aa_not_rna_top_seqs\n",
    "\n",
    "aa_list =[gly_aa_not_rna_top_seqs, leu_aa_not_rna_top_seqs, phe_aa_not_rna_top_seqs, ala_aa_not_rna_top_seqs]\n",
    "for aa_not_rna_top_seqs in aa_list:\n",
    "    print('AA'+str(1+aa_list.index(aa_not_rna_top_seqs)))\n",
    "    sequences_enrichments = []\n",
    "    counter=0\n",
    "    for seq in aa_not_rna_top_seqs:\n",
    "        counter+=1\n",
    "        if counter%1000==0:\n",
    "            print(str(counter)+'/'+str(len(aa_not_rna_top_seqs)))\n",
    "        first_6 = seq[0][5:11]\n",
    "        second_6 = seq[0][11:17]\n",
    "        enrichment_first_6 = ctrl_8a_subseqs.count(first_6)/expected_8a\n",
    "        enrichment_second_6 = ctrl_5b_subseqs.count(second_6)/expected_5b\n",
    "        enrichment = round(enrichment_first_6*enrichment_second_6,2)\n",
    "        sequences_enrichments.append([seq[0], seq[1], enrichment, seq[1]/enrichment])\n",
    "    sorted_seqs = sorted(sequences_enrichments, key=lambda x: x[-1])[::-1]\n",
    "    rscape_format('AA'+str(1+aa_list.index(aa_not_rna_top_seqs)), 1000, sorted_seqs)\n",
    "    print(sorted_seqs[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6ce854f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
